### PR TITLE
[Extend] Added default LanguageCode setting

### PR DESCRIPTION
**Pull request by AlexanderKanakis**
Preview of app setting select:
image

### CODE REVIEW

**Comment by ear-dev** [comment]
@AlexanderKanakis Did you reorder the admin options in this PR purposefully?  I kind of liked when the ES/CX swtiches and configs were at the top.

image

**Comment by AlexanderKanakis** [comment]
@ear-dev I assume that at some point you run an older version of the app where this settings did not exist. If the app settings exist in the database, the newer app settings will always be at the bottom. You can access your database, delete your app settings values and re-deploy the app.

**Comment by ear-dev** [comment]
@AlexanderKanakis does this feature work in ES and CX, or is it specific to CX?  How do you test?  I'm not sure that our agents are configured for Portuguese yet, not sure....  Is it that we are sending an attribute in a payload back to the agent?

**Comment by AlexanderKanakis** [comment]
This feature works with both ES and CX. 

When adding a new language to the agent a new instance of the flow is generated with the same structure but empty text responses in the fulfillments (as those need to change for the new language). 
image

The agent, after checking the language code in the request's query input 
```
        const queryInput = {
            ...requestType === DialogflowRequestType.EVENT && { event: { event: typeof request === 'string' ? request : request.name} },
            ...requestType === DialogflowRequestType.MESSAGE && { text: { text: request }},
            languageCode: data.custom_languageCode || defaultLanguageCode || LanguageCode.EN,
        };
```
it redirects to the correct instance of the bot.

`data.custom_languageCode` is a visitor-specified language code and it's usually undefined unless the visitor wants to change their language (see https://github.com/WideChat/Apps.Dialogflow/pull/78). So it all depends on the admin app setting `defaultLanguageCode `.

**Comment by ear-dev** [comment]
@AlexanderKanakis this looks good.....  working.  Thanks.  @Shailesh351 do you think you have time to do a quick review on this one?

File: handler/OnAgentAssignedHandler.ts [hunk 1]
```diff
@@ -52,9 +54,18 @@ export class OnAgentAssignedHandler {
 
         await updateRoomCustomFields(rid, { welcomeEventSent: true }, this.read, this.modify);
 
+        const assoc = new RocketChatAssociationRecord(RocketChatAssociationModel.MISC, `SFLAIA-${rid}`);
```
Comment (Shailesh351) [inline handler/OnAgentAssignedHandler.ts:57]: We should create a common Utility function that takes roomId as an argument and returns the `RocketChatAssociationRecord` object. And reuse that across the app.

And we should rename `SFLAIA-` to something more meaningful.

File: config/Settings.ts [hunk 2]
```diff
@@ -122,6 +123,18 @@ export const settings: Array<ISetting> = [
         i18nDescription: 'dialogflow_cx_region_desc',
         required: false,
     },
+    {
+        id: AppSetting.DialogflowDefaultLanguage,
+        public: true,
+        type: SettingType.SELECT,
+        packageValue: 'en',
+        i18nLabel: 'agent_version',
```
Comment (Shailesh351) [inline config/Settings.ts:131]: There is no entry for `agent_version` in localization files. Adding a proper description of this setting will help the admin.

File: handler/OnAgentAssignedHandler.ts [hunk 3]
```diff
@@ -52,9 +54,18 @@ export class OnAgentAssignedHandler {
...
+        const data = await retrieveDataByAssociation(this.read, assoc);
+
+        const defaultLanguageCode = await getAppSettingValue(this.read, AppSetting.DialogflowDefaultLanguage);
+
         try {
-            const event = { name: 'Welcome', languageCode: 'en', parameters: {...livechatData, roomId: rid, visitorToken} || {} };
-            const response: IDialogflowMessage = await Dialogflow.sendRequest(this.http, this.read, this.modify, this.persis, rid, event, DialogflowRequestType.EVENT);
+            const event = {
+                name: 'Welcome',
+                languageCode: data.custom_languageCode || defaultLanguageCode || 'en',
```
Comment (Shailesh351) [inline handler/OnAgentAssignedHandler.ts:65]: We should use `LanguageCode.EN` instead of `en`

**Review by bhardwajaditya — APPROVED** [review]

**Review by bhardwajaditya — COMMENTED** [review]
PR looks good apart from the changes @Shailesh351 suggested. I share the same view on that.

**Comment by ear-dev** [comment]
@Shailesh351 @bhardwajaditya It looks like Alexander is still in the hospital and taking some time to recover.  I suspect he may be out for another week or so at least.  I know it's not our usual process.......  but Shailesh's suggestions in his review were good....  do you think one of you can pull this branch and make those changes?  The reason I ask is that we would love to get the cx_development branch merged into master because it fixes a bunch of things and gets us closer to our brazil deployment. And this is the last PR we will have before the merge. Let me know if you think you have time.  Thanks!

**Comment by Shailesh351** [comment]
@ear-dev Will look into it as soon as I'm done with catch up merge 👍

**Comment by ear-dev** [comment]
@Shailesh351 what's the status on this one?  are we able to merge this now?  Thanks...

**Comment by ear-dev** [comment]
Thanks @Shailesh351

**Pull request merged**